File: backend/internal/handlers/equipments.go
package handlers

import (
	"fmt"
	"strconv"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"

	"github.com/inovar/backend/internal/middleware"
	"github.com/inovar/backend/internal/models"
)

// ListEquipments returns equipments based on user role
func (h *Handler) ListEquipments(c *fiber.Ctx) error {
	role := middleware.GetUserRole(c)
	userID := middleware.GetUserID(c)

	clientID := c.Query("clientId")
	activeOnly := c.Query("activeOnly", "true") == "true"

	var equipments []models.Equipamento
	query := h.DB.Model(&models.Equipamento{})

	// Filter by active if requested
	if activeOnly {
		query = query.Where("active = ?", true)
	}

	switch role {
	case models.RoleCliente:
		// Client only sees their own equipments
		var cliente models.Cliente
		h.DB.Where("user_id = ?", userID).First(&cliente)
		query = query.Where("client_id = ?", cliente.ID)
	default:
		// Filter by client if provided
		if clientID != "" {
			query = query.Where("client_id = ?", clientID)
		}
	}

	query.Order("location ASC").Find(&equipments)

	return Success(c, equipments)
}

// CreateEquipmentRequest represents equipment creation payload
type CreateEquipmentRequest struct {
	ClientID           string `json:"clientId"`
	Brand              string `json:"brand"`
	Model              string `json:"model"`
	BTU                int    `json:"btu"`
	SerialNumber       string `json:"serialNumber"`
	Location           string `json:"location"`
	LastPreventiveDate string `json:"lastPreventiveDate"`
	PreventiveInterval int    `json:"preventiveInterval"`
}

// CreateEquipment creates a new equipment
func (h *Handler) CreateEquipment(c *fiber.Ctx) error {
	var req CreateEquipmentRequest
	if err := c.BodyParser(&req); err != nil {
		return BadRequest(c, "Dados inválidos")
	}

	role := middleware.GetUserRole(c)
	userID := middleware.GetUserID(c)

	// For clients, use their own client ID
	clientID := req.ClientID
	if role == models.RoleCliente {
		var cliente models.Cliente
		h.DB.Where("user_id = ?", userID).First(&cliente)
		clientID = cliente.ID
	}

	if clientID == "" {
		return BadRequest(c, "Cliente é obrigatório")
	}

	if clientID == "" {
		return BadRequest(c, "Cliente é obrigatório")
	}

	// Calculate Initial Preventive Date
	interval := req.PreventiveInterval
	if interval <= 0 {
		interval = 90 // Default fallback
		var s models.Setting
		if err := h.DB.Where("key = ?", "preventive_interval").First(&s).Error; err == nil && s.Value != "" {
			if val, err := strconv.Atoi(s.Value); err == nil {
				interval = val
			}
		}
	}

	baseDate := time.Now()
	lastPreventive, _ := ParseDateTime(req.LastPreventiveDate)
	if lastPreventive != nil {
		baseDate = *lastPreventive
	}
	nextDate := baseDate.AddDate(0, 0, interval)

	equipment := models.Equipamento{
		ID:                 uuid.New().String(),
		ClientID:           clientID,
		Brand:              req.Brand,
		Model:              req.Model,
		BTU:                req.BTU,
		SerialNumber:       req.SerialNumber,
		Location:           req.Location,
		Active:             true,
		CreatedAt:          time.Now(),
		LastPreventiveDate: lastPreventive,
		NextPreventiveDate: &nextDate,
		PreventiveInterval: req.PreventiveInterval,
	}

	if err := h.DB.Create(&equipment).Error; err != nil {
		return ServerError(c, err)
	}

	h.Hub.Broadcast("equipment:created", equipment)

	return Created(c, equipment)
}

// GetEquipment returns a specific equipment
func (h *Handler) GetEquipment(c *fiber.Ctx) error {
	id := c.Params("id")
	role := middleware.GetUserRole(c)
	userID := middleware.GetUserID(c)

	var equipment models.Equipamento

	if role == models.RoleCliente {
		// Verify ownership
		var cliente models.Cliente
		h.DB.Where("user_id = ?", userID).First(&cliente)
		if err := h.DB.Where("id = ? AND client_id = ?", id, cliente.ID).First(&equipment).Error; err != nil {
			return NotFound(c, "Equipamento não encontrado")
		}
	} else {
		if err := h.DB.First(&equipment, "id = ?", id).Error; err != nil {
			return NotFound(c, "Equipamento não encontrado")
		}
	}

	return Success(c, equipment)
}

// UpdateEquipment updates an equipment
func (h *Handler) UpdateEquipment(c *fiber.Ctx) error {
	id := c.Params("id")
	role := middleware.GetUserRole(c)
	userID := middleware.GetUserID(c)

	var equipment models.Equipamento

	// Check ownership for clients
	if role == models.RoleCliente {
		var cliente models.Cliente
		h.DB.Where("user_id = ?", userID).First(&cliente)
		if err := h.DB.Where("id = ? AND client_id = ?", id, cliente.ID).First(&equipment).Error; err != nil {
			return NotFound(c, "Equipamento não encontrado")
		}
	} else {
		if err := h.DB.First(&equipment, "id = ?", id).Error; err != nil {
			return NotFound(c, "Equipamento não encontrado")
		}
	}

	var req CreateEquipmentRequest
	if err := c.BodyParser(&req); err != nil {
		return BadRequest(c, "Dados inválidos")
	}

	before := equipment // Copy original state
	equipment.UpdatedAt = time.Now()

	h.DB.Save(&equipment)

	h.Hub.Broadcast("equipment:updated", equipment)

	// Final Audit
	h.LogAudit(c, "Equipment", equipment.ID, "UPDATE", fmt.Sprintf("Updated equipment %s (%s)", equipment.Model, equipment.Brand), before, equipment)

	return Success(c, equipment)
}

// DeactivateEquipment deactivates an equipment (never delete!)
func (h *Handler) DeactivateEquipment(c *fiber.Ctx) error {
	id := c.Params("id")

	var equipment models.Equipamento
	if err := h.DB.First(&equipment, "id = ?", id).Error; err != nil {
		return NotFound(c, "Equipamento não encontrado")
	}

	equipment.Active = false
	equipment.UpdatedAt = time.Now()
	h.DB.Save(&equipment)

	h.Hub.Broadcast("equipment:deactivated", equipment)

	return Success(c, equipment)
}

// ReactivateEquipment reactivates an equipment
func (h *Handler) ReactivateEquipment(c *fiber.Ctx) error {
	id := c.Params("id")

	var equipment models.Equipamento
	if err := h.DB.First(&equipment, "id = ?", id).Error; err != nil {
		return NotFound(c, "Equipamento não encontrado")
	}

	equipment.Active = true
	equipment.UpdatedAt = time.Now()
	h.DB.Save(&equipment)

	h.Hub.Broadcast("equipment:reactivated", equipment)

	return Success(c, equipment)
}

// DeleteEquipment permanently deletes (ADMIN ONLY!)
func (h *Handler) DeleteEquipment(c *fiber.Ctx) error {
	id := c.Params("id")

	var equipment models.Equipamento
	if err := h.DB.First(&equipment, "id = ?", id).Error; err != nil {
		return NotFound(c, "Equipamento não encontrado")
	}

	// Permanent delete - also for Prestador
	// Start transaction to clean up junction table
	tx := h.DB.Begin()

	// 1. Delete associations in solicitacao_equipamentos
	if err := tx.Where("equipamento_id = ?", id).Delete(&models.SolicitacaoEquipamento{}).Error; err != nil {
		tx.Rollback()
		return ServerError(c, err)
	}

	// 2. Delete Checklists associated with this equipment
	if err := tx.Unscoped().Where("equipamento_id = ?", id).Delete(&models.Checklist{}).Error; err != nil {
		tx.Rollback()
		return ServerError(c, err)
	}

	// 3. Delete equipment itself
	if err := tx.Unscoped().Delete(&equipment).Error; err != nil {
		tx.Rollback()
		return ServerError(c, err)
	}

	if err := tx.Commit().Error; err != nil {
		return ServerError(c, err)
	}

	h.Hub.Broadcast("equipment:deleted", fiber.Map{"id": id})

	return Success(c, fiber.Map{"message": "Equipamento excluído permanentemente"})
}

File: infra/pyservice/routes/equipments.py
"""Equipments CRUD routes — called by Go via HTTP"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional
from datetime import datetime, timedelta
import uuid

from database import get_db
from models import Equipamento

router = APIRouter()

class EquipCreate(BaseModel):
    client_id: str
    company_id: str = ""
    brand: str
    model: str
    btu: int
    serial_number: str = ""
    location: str
    preventive_interval: int = 0

class EquipUpdate(BaseModel):
    brand: Optional[str] = None
    model: Optional[str] = None
    btu: Optional[int] = None
    serial_number: Optional[str] = None
    location: Optional[str] = None
    preventive_interval: Optional[int] = None

def equip_to_dict(e: Equipamento) -> dict:
    return {
        "id": e.id, "clientId": e.client_id, "companyId": e.company_id or "",
        "brand": e.brand, "model": e.model, "btu": e.btu,
        "serialNumber": e.serial_number or "", "location": e.location,
        "preventiveInterval": e.preventive_interval or 0,
        "active": e.active,
        "lastPreventiveDate": e.last_preventive_date.isoformat() if e.last_preventive_date else None,
        "nextPreventiveDate": e.next_preventive_date.isoformat() if e.next_preventive_date else None,
        "createdAt": e.created_at.isoformat() if e.created_at else "",
    }

@router.get("")
def list_equipments(company_id: str = "", client_id: str = "", active_only: str = "", db: Session = Depends(get_db)):
    q = db.query(Equipamento).filter(Equipamento.deleted_at.is_(None))
    if company_id:
        q = q.filter(Equipamento.company_id == company_id)
    if client_id:
        q = q.filter(Equipamento.client_id == client_id)
    if active_only == "true":
        q = q.filter(Equipamento.active == True)
    return {"data": [equip_to_dict(e) for e in q.order_by(Equipamento.created_at.desc()).all()]}

@router.get("/{equip_id}")
def get_equipment(equip_id: str, db: Session = Depends(get_db)):
    e = db.query(Equipamento).filter(Equipamento.id == equip_id, Equipamento.deleted_at.is_(None)).first()
    if not e:
        raise HTTPException(404, "Equipment not found")
    return {"data": equip_to_dict(e)}

@router.post("", status_code=201)
def create_equipment(req: EquipCreate, db: Session = Depends(get_db)):
    now = datetime.utcnow()
    next_prev = None
    if req.preventive_interval > 0:
        next_prev = now + timedelta(days=req.preventive_interval)

    equip = Equipamento(
        id=str(uuid.uuid4()), client_id=req.client_id, company_id=req.company_id,
        brand=req.brand, model=req.model, btu=req.btu,
        serial_number=req.serial_number, location=req.location,
        preventive_interval=req.preventive_interval,
        last_preventive_date=now, next_preventive_date=next_prev,
        active=True, created_at=now,
    )
    db.add(equip)
    db.commit()
    db.refresh(equip)
    return {"data": equip_to_dict(equip)}

@router.put("/{equip_id}")
def update_equipment(equip_id: str, req: EquipUpdate, db: Session = Depends(get_db)):
    e = db.query(Equipamento).filter(Equipamento.id == equip_id, Equipamento.deleted_at.is_(None)).first()
    if not e:
        raise HTTPException(404, "Equipment not found")
    for k, v in req.model_dump(exclude_none=True).items():
        setattr(e, k, v)
    e.updated_at = datetime.utcnow()
    db.commit()
    db.refresh(e)
    return {"data": equip_to_dict(e)}

@router.patch("/{equip_id}/deactivate")
def deactivate(equip_id: str, db: Session = Depends(get_db)):
    e = db.query(Equipamento).filter(Equipamento.id == equip_id).first()
    if not e:
        raise HTTPException(404, "Equipment not found")
    e.active = False
    e.updated_at = datetime.utcnow()
    db.commit()
    return {"data": equip_to_dict(e)}

@router.patch("/{equip_id}/reactivate")
def reactivate(equip_id: str, db: Session = Depends(get_db)):
    e = db.query(Equipamento).filter(Equipamento.id == equip_id).first()
    if not e:
        raise HTTPException(404, "Equipment not found")
    e.active = True
    e.updated_at = datetime.utcnow()
    db.commit()
    return {"data": equip_to_dict(e)}

@router.delete("/{equip_id}")
def delete_equipment(equip_id: str, db: Session = Depends(get_db)):
    e = db.query(Equipamento).filter(Equipamento.id == equip_id).first()
    if not e:
        raise HTTPException(404, "Equipment not found")
    e.deleted_at = datetime.utcnow()
    db.commit()
    return {"message": "Deleted"}
